runtime/autoload/ccomplete.lua

```diff
@@ -0,0 +1,892 @@
+    -- # TODO: join previous line if it makes sense
+    local line = NVIM9.fn['getline']('.')
+    local col = NVIM9.fn['charcol']('.')
+    if NVIM9.fn['stridx'](NVIM9.slice(line, nil, NVIM9.ops['Minus'](col, 1)), ';') >= 0 then
+      -- # Handle multiple declarations on the same line.
+      local col2 = NVIM9.ops['Minus'](col, 1)
+      while NVIM9.index(line, col2) ~= ';' do
+        col2 = col2 - 1
+      end
+      line = NVIM9.slice(line, NVIM9.ops['Plus'](col2, 1), nil)
+      col = NVIM9.ops['Minus'](col, col2)
+    end
+    if NVIM9.fn['stridx'](NVIM9.slice(line, nil, NVIM9.ops['Minus'](col, 1)), ',') >= 0 then
+      -- # Handle multiple declarations on the same line in a function
+      -- # declaration.
+      local col2 = NVIM9.ops['Minus'](col, 1)
+      while NVIM9.index(line, col2) ~= ',' do
+        col2 = col2 - 1
+      end
+      if
+        NVIM9.ops['RegexpMatches'](
+          NVIM9.slice(line, NVIM9.ops['Plus'](col2, 1), NVIM9.ops['Minus'](col, 1)),
+          ' *[^ ][^ ]*  *[^ ]'
+        )
+      then
+        line = NVIM9.slice(line, NVIM9.ops['Plus'](col2, 1), nil)
+        col = NVIM9.ops['Minus'](col, col2)
+      end
+    end
+    if NVIM9.fn['len'](items) == 1 then
+      -- # Completing one word and it's a local variable: May add '[', '.' or
+      -- # '->'.
+      local match = NVIM9.index(items, 0)
+      local kind = 'v'
+      if NVIM9.fn['match'](line, '\\<' .. match .. '\\s*\\[') > 0 then
+        match = match .. '['
+      else
+        res = Nextitem(NVIM9.slice(line, nil, NVIM9.ops['Minus'](col, 1)), { '' }, 0, true)
+        if NVIM9.fn['len'](res) > 0 then
+          -- # There are members, thus add "." or "->".
+          if NVIM9.fn['match'](line, '\\*[ \\t(]*' .. match .. '\\>') > 0 then
+            match = match .. '->'
+          else
+            match = match .. '.'
+          end
+        end
+      end
+      res = { { ['match'] = match, ['tagline'] = '', ['kind'] = kind, ['info'] = line } }
+    elseif NVIM9.bool(NVIM9.fn['len'](items) == NVIM9.ops['Plus'](arrays, 1)) then
+      -- # Completing one word and it's a local array variable: build tagline
+      -- # from declaration line
+      local match = NVIM9.index(items, 0)
+      local kind = 'v'
+      local tagline = '\t/^' .. line .. '$/'
+      res = { { ['match'] = match, ['tagline'] = tagline, ['kind'] = kind, ['info'] = line } }
+    else
+      -- # Completing "var.", "var.something", etc.
+      res = Nextitem(
+        NVIM9.slice(line, nil, NVIM9.ops['Minus'](col, 1)),
+        NVIM9.slice(items, 1, nil),
+        0,
+        true
+      )
+    end
+  end
+
+  if NVIM9.fn['len'](items) == 1 or NVIM9.fn['len'](items) == NVIM9.ops['Plus'](arrays, 1) then
+    -- # Only one part, no "." or "->": complete from tags file.
+    local tags = {}
+    if NVIM9.fn['len'](items) == 1 then
+      tags = NVIM9.fn['taglist']('^' .. base)
+    else
+      tags = NVIM9.fn['taglist']('^' .. NVIM9.index(items, 0) .. '$')
```

Comment:
is it much trouble to use sugar when possible?
```suggestion
      tags = NVIM9.fn.taglist('^' .. NVIM9.index(items, 0) .. '$')
```

Reply:
Done.

Reply:
(handles illegal names as well and moves it back to strings as well, btw)